lib/Sema/TypeCheckGeneric.cpp

```diff
@@ -231,6 +267,67 @@ CompleteGenericTypeResolver::recordParamType(ParamDecl *decl, Type type) {
   decl->setInterfaceType(type);
 }
 
+bool TypeChecker::validateRequirement(SourceLoc whereLoc, RequirementRepr &req,
+                                      DeclContext *lookupDC,
+                                      TypeResolutionOptions options,
+                                      GenericTypeResolver *resolver) {
+  if (req.isInvalid())
+    return true;
+
+  switch (req.getKind()) {
+  case RequirementReprKind::TypeConstraint: {
+    // Validate the types.
+    if (validateType(req.getSubjectLoc(), lookupDC, options, resolver)) {
+      req.setInvalid();
+      return true;
+    }
+
+    if (validateType(req.getConstraintLoc(), lookupDC, options, resolver)) {
+      req.setInvalid();
+      return true;
+    }
+
+    // FIXME: Feels too early to perform this check.
```

Comment:
I know this isn't your FIXME; this should eventually be handled by the generic signature builder, I think.